subsys/testsuite/include/negative_handler.h

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr.h>
+
+
+#if defined(CONFIG_NEGATIVE_TEST_FATAL_HANDLER)
+/* These valid flag used to figure out if the fault was expected or not. */
+extern volatile bool valid_fault;
+extern void *fatal_hook_param;
+
+/**
+ * @brief Set the flag that treat fatal error happened as expected
+ *
+ * This is used for negative test cases which triggers a fatal error.
+ * Set the param true will still pass the test case when expected fatal
+ * error happened. For normal test case, set it false makes it back to
+ * normal behavior.
+ *
+ * @param valid flag indicate fault is expected
+ */
+static inline void set_fault_valid(bool valid)
+{
+	valid_fault = valid;
+	/* Put a barrier here, such that no instructions get ordered by the
+	 * compiler before we set valid_fault. This can happen with expansion
+	 * of inline syscall invocation functions.
+	 */
+	compiler_barrier();
+}
+
+/**
+ * @brief Passing the parameter from test case to fatal error handler hook
+ *
+ * Call this function to pass parameter to predefined fatal handler, it makes
+ * us taking different actions in our test case (applications).
+ */
+static inline void set_fatal_hook_param(void *param)
+{
+	fatal_hook_param = param;
+}
+#endif
+
+
+#if defined(CONFIG_NEGATIVE_TEST_ASSERT_HANDLER)
+/* These valid flag used to figure out if the assert was expected or not. */
+extern volatile bool valid_assert;
+extern void *assert_hook_param;
+
+/**
+ * @brief Set the flag that treat assert fail happened as expected
+ *
+ * This is used for negative test cases which triggers a assert fail.
+ * Set the param true will still pass the test case when expected assert
+ * fail happened. For normal test case, set it false make it back to
+ * normal behavior.
+ *
+ * @param valid flag indicate assert is expected
+ */
+static inline void set_assert_valid(bool valid)
+{
+	valid_assert = valid;
+	/* Put a barrier here, such that no instructions get ordered by the
+	 * compiler before we set valid_fault. This can happen with expansion
+	 * of inline syscall invocation functions.
+	 */
+	compiler_barrier();
```

Comment:
"valid_assert" is already marked volatile.  The comment and compiler barrier are needless, preventing reordering of access is exactly what "volatile" means.  Remove.

Reply:
Got it, so glad to learn this concept and will study it ! I saw the code in tests/kernel/mem_protect/mem_protect/src/mem_protect.h using it this way, so I refer to it.  I will update them, thanks!